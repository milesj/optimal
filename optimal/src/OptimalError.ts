import { ValidationError } from './ValidationError';

export class OptimalError extends Error {
	errors: ValidationError[] = [];

	file: string = '';

	schema: string = '';

	constructor() {
		super('');

		this.name = 'OptimalError';
	}
}
