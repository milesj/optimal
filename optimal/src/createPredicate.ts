import { AnySchema, InferSchemaType, Predicate } from './types';

/**
 * Create a reusable predicate from a schema that validates a value to
 * return a boolean if no validation errors arise. If null or undefined
 * is returned, will return `false`, regardless of nullable or undefinable
 * state.
 */
export function createPredicate<S extends AnySchema, T = InferSchemaType<S>>(
	schema: S,
): Predicate<T> {
	return (value) => {
		try {
			if (value == null) {
				throw new Error('Avoid null/undefined');
			}

			schema.validate(value);
		} catch {
			return false;
		}

		return true;
	};
}
