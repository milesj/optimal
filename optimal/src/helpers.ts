import {
	Constructor,
	DefaultValue,
	DefaultValueInitializer,
	Schema,
	SchemaValidateOptions,
	UnknownObject,
} from './types';
import { ValidationError } from './ValidationError';

export function isObject(value: unknown): value is object {
	return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
	return (
		isObject(value) &&
		typeof (value as UnknownObject).schema === 'function' &&
		typeof (value as UnknownObject).state === 'function' &&
		typeof (value as UnknownObject).type === 'function' &&
		typeof (value as UnknownObject).validate === 'function'
	);
}

export function isValidDate(value: Date): value is Date {
	return (
		Object.prototype.toString.call(value) === '[object Date]' &&
		!Number.isNaN(value.getTime()) &&
		value.toString() !== 'Invalid Date'
	);
}

export function isValidNumber(value: unknown): value is number {
	return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
	return typeof value === 'string' && value !== '';
}

export function createArray(value: unknown): unknown[] {
	if (value === undefined) {
		return [];
	}

	return Array.isArray(value) ? [...(value as unknown[])] : [value];
}

export function createDate(value: unknown): Date {
	if (value instanceof Date) {
		return value;
	}

	if (value === undefined || value === null) {
		return new Date();
	}

	return new Date(value as number);
}

export function createObject<T = UnknownObject>(value: unknown) {
	return (isObject(value) ? { ...value } : {}) as unknown as T;
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(object: unknown, contract: Constructor<T>): object is T {
	if (!object || typeof object !== 'object') {
		return false;
	}

	if (object instanceof contract) {
		return true;
	}

	let current = object;

	while (current) {
		if (current.constructor.name === 'Object') {
			return false;
		}

		if (
			current.constructor.name === contract.name ||
			(current instanceof Error && current.name === contract.name)
		) {
			return true;
		}

		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		current = Object.getPrototypeOf(current);
	}

	return false;
}

export function invalid(
	condition: boolean,
	message: string,
	path: string = '',
	value: unknown = undefined,
) {
	if (condition) {
		return;
	}

	throw new ValidationError(message, path, value);
}

export function invariant(condition: boolean, message: string) {
	if (condition) {
		return;
	}

	throw new Error(message);
}

export function logUnknown(unknownFields: object, pathPrefix?: string) {
	const unknownKeys = Object.keys(unknownFields);

	if (unknownKeys.length > 0) {
		throw new Error(
			`${pathPrefix ? `Unknown "${pathPrefix}" fields` : 'Unknown fields'}: ${unknownKeys.join(
				', ',
			)}.`,
		);
	}
}

export function pathKey(path: string): string {
	if (path.endsWith(']')) {
		const index = path.lastIndexOf('[');

		return index > 0 ? path.slice(index) : path;
	}

	const index = path.lastIndexOf('.');

	return index > 0 ? path.slice(index + 1) : path;
}

export function prettyValue(value: unknown): string | null {
	switch (typeof value) {
		case 'string':
			return `"${value}"`;

		case 'number':
		case 'function':
			return String(value);

		case 'object': {
			if (value === null) {
				return `\`null\``;
			}

			if (value.constructor !== Object) {
				return value.constructor.name === 'Array' ? null : `\`${value.constructor.name}\``;
			}

			return null;
		}

		default:
			return `\`${value}\``;
	}
}

export function tryAndCollect(
	validator: () => boolean | void,
	validError: ValidationError,
	collectErrors?: boolean,
): boolean {
	let result = false;

	try {
		const value = validator();

		if (typeof value === 'boolean') {
			result = value;
		}
	} catch (error: unknown) {
		if (error instanceof Error && collectErrors) {
			validError.addError(error);
		} else {
			throw error;
		}
	}

	return result;
}

export function extractDefaultValue<T>(
	defaultValue: DefaultValue<T>,
	path: string,
	{ currentObject, rootObject }: SchemaValidateOptions,
) {
	return typeof defaultValue === 'function'
		? (defaultValue as DefaultValueInitializer<T>)(path, currentObject!, rootObject!)
		: defaultValue;
}
