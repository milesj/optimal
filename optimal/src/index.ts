/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import * as schemas from './schemas';

export type Schemas = typeof schemas;

export { schemas };
export * from './createPredicate';
export * from './createSchema';
export * from './criteria';
export * from './optimal';
export * from './OptimalError';
export * from './schemas/array';
export * from './schemas/blueprint';
export * from './schemas/bool';
export * from './schemas/custom';
export * from './schemas/date';
export * from './schemas/func';
export * from './schemas/instance';
export * from './schemas/lazy';
export * from './schemas/number';
export * from './schemas/object';
export * from './schemas/regex';
export * from './schemas/schema';
export * from './schemas/shape';
export * from './schemas/string';
export * from './schemas/tuple';
export * from './schemas/union';
export * from './types';
export * from './ValidationError';
