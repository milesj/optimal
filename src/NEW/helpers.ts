import { Constructor, UnknownObject, Schema } from './types';

export function createDate(value: unknown): Date {
  return value instanceof Date ? value : new Date(value as number);
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(
  object: unknown,
  contract: Constructor<T> | Function,
): object is T {
  if (!object || typeof object !== 'object') {
    return false;
  }

  if (object instanceof contract) {
    return true;
  }

  let current = object;

  while (current) {
    if (current.constructor.name === 'Object') {
      return false;
    }

    if (
      current.constructor.name === contract.name ||
      (current instanceof Error && current.name === contract.name)
    ) {
      return true;
    }

    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    current = Object.getPrototypeOf(current);
  }

  return false;
}

export function invariant(condition: boolean, message: string, path: string = '') {
  if (__DEV__) {
    if (condition) {
      return;
    }

    // TODO no errorPrefix
    const prefix = path ? `Invalid field "${path}".` : '';

    throw new Error(`${prefix} ${message}`.trim());
  }
}

export function isObject(value: unknown): value is object {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
  return (
    isObject(value) &&
    typeof (value as UnknownObject).validate === 'function' &&
    typeof (value as UnknownObject).type === 'function'
  );
}

export function isValidDate(value: Date): value is Date {
  return (
    Object.prototype.toString.call(value) === '[object Date]' &&
    !Number.isNaN(value.getTime()) &&
    value.toString() !== 'Invalid Date'
  );
}

export function isValidNumber(value: unknown): value is number {
  return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
  return typeof value === 'string' && value !== '';
}

export function pathKey(path: string): string {
  const index = path.lastIndexOf('.');

  return index > 0 ? path.slice(index + 1) : path;
}
