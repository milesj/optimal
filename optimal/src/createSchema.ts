import { invalid, tryAndCollect } from './helpers';
import { OptimalError } from './OptimalError';
import {
	AnySchema,
	Criteria,
	CriteriaFactory,
	DefaultValueInitializer,
	InferSchemaType,
	Schema,
	SchemaOptions,
	SchemaState,
	SchemaValidateOptions,
} from './types';

/**
 * Run all validation checks that have been enqueued and return a type casted value.
 * If a value is undefined, inherit the default value, else throw if required.
 * If nullable and the value is null, return early.
 */
function validate<T>(
	state: SchemaState<T>,
	validators: Criteria<T>[],
	initialValue: unknown,
	path: string = '',
	{
		collectErrors = true,
		currentObject = {},
		rootObject = currentObject,
	}: SchemaValidateOptions = {},
): T | null {
	const { defaultValue, metadata } = state;

	let value: unknown = initialValue;

	// Handle undefined
	if (value === undefined) {
		value =
			typeof defaultValue === 'function'
				? (defaultValue as DefaultValueInitializer<T>)(path, currentObject, rootObject)
				: defaultValue;

		invalid(!state.required, 'Field is required and must be defined.', path);
	} else {
		if (__DEV__ && metadata.deprecatedMessage) {
			// eslint-disable-next-line no-console
			console.info(`Field "${path}" is deprecated. ${metadata.deprecatedMessage}`);
		}

		invalid(!state.never, 'Field should never be used.', path);
	}

	// Handle null
	if (value === null) {
		invalid(state.nullable, 'Null is not allowed.', path);
	}

	// Run validations and produce a new value
	const optimalError = new OptimalError();

	validators.forEach((test) => {
		if (
			(test.skipIfNull && value === null) ||
			(test.skipIfOptional && !state.required && value === state.defaultValue)
		) {
			return;
		}

		tryAndCollect(
			() => {
				const result = test.validate(value as T, path, {
					collectErrors,
					currentObject,
					rootObject,
				});

				if (result !== undefined) {
					value = result as T;
				}
			},
			optimalError,
			collectErrors,
		);
	});

	if (optimalError.errors.length > 0) {
		throw optimalError;
	}

	return value as T;
}

export function createSchema<S extends AnySchema, T = InferSchemaType<S>>(
	{ api, cast, defaultValue, type }: SchemaOptions<T>,
	criteria: (Criteria<T> | CriteriaFactory<T>)[] = [],
): S {
	const state: SchemaState<T> = {
		defaultValue,
		metadata: {},
		never: false,
		nullable: false,
		required: false,
		type,
	};

	const validators: Criteria<T>[] = [];

	criteria.forEach((crit) => {
		if (typeof crit === 'function') {
			const validator = crit(state);

			if (validator) {
				validators.push(validator);
			}
		} else {
			validators.push(crit);
		}
	});

	const schema: Schema<T> = {
		schema() {
			return type;
		},
		type() {
			return state.type;
		},
		validate(value, path, options) {
			const result = validate(state, validators, value, path, options)!;

			return cast && result !== null ? cast(result) : result;
		},
	};

	Object.entries(api).forEach(([name, method]) => {
		Object.defineProperty(schema, name, {
			enumerable: true,
			// Create a new schema so that our chainable API is immutable
			value: (...args: unknown[]) =>
				createSchema({ api, cast, defaultValue, type }, [
					...criteria,
					(nextState) => method(nextState, ...args),
				]),
		});
	});

	return schema as S;
}
