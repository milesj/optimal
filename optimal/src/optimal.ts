import { isObject } from './helpers';
import { OptimalError } from './OptimalError';
import { shape } from './schemas/shape';
import { Blueprint, DeepPartial, SchemaValidateOptions, UnknownObject } from './types';
import { ValidationError } from './ValidationError';

export interface OptimalOptions {
	file?: string;
	name?: string;
	prefix?: string;
	unknown?: boolean;
}

export interface Optimal<T extends object> {
	configure: (options: OptimalOptions) => void;
	validate: (struct: DeepPartial<T>, options?: SchemaValidateOptions) => Required<T>;
}

export function optimal<Schemas extends object>(
	blueprint: Blueprint<Schemas>,
	baseOpts: OptimalOptions = {},
): Optimal<Schemas> {
	const options: OptimalOptions = {};
	let schema = shape(blueprint);

	function configure(nextOpts: OptimalOptions) {
		if (!isObject(nextOpts)) {
			throw new TypeError('Optimal options must be a plain object.');
		}

		Object.assign(options, nextOpts);
		schema = schema.exact(!options.unknown);
	}

	configure(baseOpts);

	return {
		configure,
		validate(struct, validateOptions) {
			const object = struct as UnknownObject;

			try {
				return schema.validate(struct, options.prefix ?? '', {
					...validateOptions,
					currentObject: object,
					rootObject: object,
				});
			} catch (error: unknown) {
				let invalid: OptimalError;
				let prefix = '';

				if (error instanceof OptimalError) {
					invalid = error;
				} else {
					invalid = new OptimalError();
					invalid.addError(error as ValidationError);
				}

				if (options.name) {
					invalid.schema = options.name;
					prefix = options.name;
				}

				if (options.file) {
					invalid.file = options.file;
					prefix = prefix ? `${prefix} (${options.file})` : options.file;
				}

				if (prefix) {
					invalid.message = `${prefix}: ${invalid.message}`;
				}

				throw invalid;
			}
		},
	};
}
