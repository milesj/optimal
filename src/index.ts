/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import optimal from './optimal';
import Schema from './Schema';
import Predicate, { custom, func } from './Predicate';
import ArrayPredicate, { array } from './predicates/Array';
import BooleanPredicate, { bool } from './predicates/Boolean';
import InstancePredicate, { instance, date, predicate, regex } from './predicates/Instance';
import NumberPredicate, { number } from './predicates/Number';
import ObjectPredicate, { object, blueprint } from './predicates/Object';
import ShapePredicate, { shape } from './predicates/Shape';
import StringPredicate, { string } from './predicates/String';
import TuplePredicate, { tuple } from './predicates/Tuple';
import UnionPredicate, { union } from './predicates/Union';

export const predicates = {
  array,
  blueprint,
  bool,
  custom,
  date,
  func,
  instance,
  number,
  object,
  predicate,
  regex,
  shape,
  string,
  tuple,
  union,
};

export {
  array,
  blueprint,
  bool,
  custom,
  date,
  func,
  instance,
  number,
  object,
  predicate,
  regex,
  shape,
  string,
  tuple,
  union,
};

export {
  Schema,
  Predicate,
  ArrayPredicate,
  BooleanPredicate,
  InstancePredicate,
  NumberPredicate,
  ObjectPredicate,
  ShapePredicate,
  StringPredicate,
  TuplePredicate,
  UnionPredicate,
};

export type Predicates = typeof predicates;

export * from './types';

export default optimal;
