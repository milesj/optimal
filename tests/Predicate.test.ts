import { custom, func, Schema, Predicate } from '../src';
import { runChecks } from './helpers';

describe('Predicate', () => {
  let schema: Schema<{}>;
  let predicate: Predicate<unknown>;

  beforeEach(() => {
    schema = new Schema({});
    predicate = new Predicate('string', 'foo');
    predicate.schema = schema;
  });

  it('errors if default value is undefined', () => {
    expect(() => {
      predicate = new Predicate('string', undefined);
    }).toThrowErrorMatchingSnapshot();
  });

  it('sets the type and default value', () => {
    expect(predicate.type).toBe('string');
    expect(predicate.defaultValue).toBe('foo');
  });

  describe('and()', () => {
    beforeEach(() => {
      predicate.and('bar', 'baz');
    });

    it('errors if no keys are defined', () => {
      expect(() => {
        predicate.and();
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if not all options are defined', () => {
      expect(() => {
        runChecks(predicate, 'a', {
          key: 'foo',
          struct: {
            foo: 'a',
            baz: 'c',
          },
        });
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if all are defined', () => {
      expect(() => {
        runChecks(predicate, 'a', {
          key: 'foo',
          struct: {
            foo: 'a',
            bar: 'b',
            baz: 'c',
          },
        });
      }).not.toThrow();
    });
  });

  describe('checkType()', () => {
    describe('array', () => {
      it('allows arrays', () => {
        predicate.type = 'array';

        expect(() => {
          runChecks(predicate, []);
        }).not.toThrow();
      });

      it('errors on non-arrays', () => {
        predicate.type = 'array';

        expect(() => {
          runChecks(predicate, 123);
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('boolean', () => {
      it('allows booleans', () => {
        predicate.type = 'boolean';

        expect(() => {
          runChecks(predicate, true);
        }).not.toThrow();
      });

      it('errors on non-booleans', () => {
        predicate.type = 'boolean';

        expect(() => {
          runChecks(predicate, 123);
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('function', () => {
      it('allows functions', () => {
        predicate.type = 'function';

        expect(() => {
          runChecks(predicate, () => {});
        }).not.toThrow();
      });

      it('errors on non-functions', () => {
        predicate.type = 'function';

        expect(() => {
          runChecks(predicate, 'foo');
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('number', () => {
      it('allows numbers', () => {
        predicate.type = 'number';

        expect(() => {
          runChecks(predicate, 123);
        }).not.toThrow();
      });

      it('errors on non-numbers', () => {
        predicate.type = 'number';

        expect(() => {
          runChecks(predicate, 'foo');
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('object', () => {
      it('allows objects', () => {
        predicate.type = 'object';

        expect(() => {
          runChecks(predicate, {});
        }).not.toThrow();
      });

      it('errors on non-objects', () => {
        predicate.type = 'object';

        expect(() => {
          runChecks(predicate, 123);
        }).toThrowErrorMatchingSnapshot();
      });

      it('errors on arrays', () => {
        predicate.type = 'object';

        expect(() => {
          runChecks(predicate, []);
        }).toThrowErrorMatchingSnapshot();
      });

      it('errors on nulls', () => {
        predicate.type = 'object';

        expect(() => {
          runChecks(predicate, null);
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('string', () => {
      it('allows strings', () => {
        predicate.type = 'string';

        expect(() => {
          runChecks(predicate, 'foo');
        }).not.toThrow();
      });

      it('errors on non-strings', () => {
        predicate.type = 'string';

        expect(() => {
          runChecks(predicate, 123);
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe('custom()', () => {
    it('errors if no callback', () => {
      expect(() =>
        // @ts-ignore
        predicate.custom(),
      ).toThrow('Custom blueprints require a validation function.');
    });

    it('errors if callback is not a function', () => {
      expect(() =>
        // @ts-ignore
        predicate.custom(123),
      ).toThrow('Custom blueprints require a validation function.');
    });

    it('triggers callback function', () => {
      predicate = custom(value => {
        if (value === 123) {
          throw new Error('This will error!');
        }
      }, 0);

      expect(() => {
        runChecks(predicate, 123);
      }).toThrowErrorMatchingSnapshot();

      expect(() => {
        runChecks(predicate, 456);
      }).not.toThrow();
    });

    it('is passed entire options object', () => {
      predicate = custom<unknown, { foo?: number; bar?: number }>((value, s) => {
        if (s.struct.foo && s.struct.bar) {
          throw new Error('This will error!');
        }
      }, 0);

      expect(() => {
        runChecks(predicate, 123, {
          struct: { foo: 123, bar: 456, error: '' },
        });
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('invariant()', () => {
    it('does nothing if condition is true', () => {
      expect(() => {
        predicate.invariant(true, '');
      }).not.toThrow();
    });

    it('errors if condition is false', () => {
      expect(() => {
        predicate.invariant(false, 'Failure');
      }).toThrowErrorMatchingSnapshot();
    });

    it('includes an option path', () => {
      expect(() => {
        predicate.invariant(false, 'Failure', 'foo.bar');
      }).toThrowErrorMatchingSnapshot();
    });

    it('includes a class name', () => {
      expect(() => {
        schema.setName('FooBar');

        predicate.invariant(false, 'Failure', 'foo.bar');
      }).toThrowErrorMatchingSnapshot();
    });

    it('includes a class name when no path', () => {
      expect(() => {
        schema.setName('FooBar');

        predicate.invariant(false, 'Failure');
      }).toThrowErrorMatchingSnapshot();
    });

    it('includes a file name', () => {
      expect(() => {
        schema.setFile('package.json');

        predicate.invariant(false, 'Failure', 'foo.bar');
      }).toThrowErrorMatchingSnapshot();
    });

    it('includes a file and class name', () => {
      expect(() => {
        schema.setName('FooBar').setFile('package.json');

        predicate.invariant(false, 'Failure', 'foo.bar');
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('key()', () => {
    it('returns as-is if not deep', () => {
      // @ts-ignore Allow access
      expect(predicate.key('foo')).toBe('foo');
    });

    it('returns last part of a deep path', () => {
      // @ts-ignore Allow access
      expect(predicate.key('foo.bar.baz')).toBe('baz');
    });
  });

  describe('message()', () => {
    it('errors for empty value', () => {
      expect(() => {
        predicate.message('');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors for non-string value', () => {
      expect(() => {
        // @ts-ignore
        predicate.message(123);
      }).toThrowErrorMatchingSnapshot();
    });

    it('sets message', () => {
      predicate.message('foobar');

      // @ts-ignore Allow access
      expect(predicate.errorMessage).toBe('foobar');
    });
  });

  describe('deprecate()', () => {
    it('errors for empty value', () => {
      expect(() => {
        predicate.deprecate('');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors for non-string value', () => {
      expect(() => {
        // @ts-ignore
        predicate.deprecate(123);
      }).toThrowErrorMatchingSnapshot();
    });

    it('sets message', () => {
      predicate.deprecate('foobar');

      // @ts-ignore Allow access
      expect(predicate.deprecatedMessage).toBe('foobar');
    });
  });

  describe('nullable()', () => {
    it('errors if field is undefined', () => {
      expect(() => {
        predicate.notNullable();

        runChecks(predicate, null);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if field is defined', () => {
      expect(() => {
        predicate.nullable();

        runChecks(predicate, null);
      }).not.toThrow();
    });
  });

  describe('required()', () => {
    it('errors if field is undefined', () => {
      expect(() => {
        predicate.required();

        runChecks(predicate);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if field is defined', () => {
      expect(() => {
        predicate.required(false);

        runChecks(predicate, 'foo');
      }).not.toThrow();
    });
  });

  describe('run()', () => {
    it('returns valid value', () => {
      expect(runChecks(predicate, 'bar')).toBe('bar');
    });

    it('returns default value if value passed is undefined', () => {
      expect(runChecks(predicate)).toBe('foo');
    });

    it('returns null if value passed is null and predicate is nullable', () => {
      predicate.nullable();

      expect(runChecks(predicate, null)).toBeNull();
    });

    it('errors if value passed is undefined and predicate is required', () => {
      predicate.required();

      expect(() => {
        runChecks(predicate);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value passed is null and predicate is non-nullable', () => {
      expect(() => {
        runChecks(predicate, null);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value passed is defined and predicate is never', () => {
      predicate.never();

      expect(() => {
        expect(runChecks(predicate, 'bar')).toBeUndefined();
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs default type of check', () => {
      expect(() => {
        runChecks(predicate, 123);
      }).toThrowErrorMatchingSnapshot();
    });

    it('uses custom message', () => {
      predicate.message('Oops, something is broken.');

      expect(() => {
        runChecks(predicate, 123);
      }).toThrowErrorMatchingSnapshot();
    });

    describe('deprecation', () => {
      const oldInfo = console.info;

      beforeEach(() => {
        jest.spyOn(console, 'info').mockImplementation();
      });

      afterEach(() => {
        console.info = oldInfo;
      });

      it('logs a message', () => {
        predicate.deprecate('Use something else.');
        runChecks(predicate, 'foo');

        expect(console.info).toHaveBeenCalledWith('Field "key" is deprecated. Use something else.');
      });

      it('doesnt log if undefined', () => {
        predicate.deprecate('Use something else.');
        runChecks(predicate);

        expect(console.info).not.toHaveBeenCalledWith(
          'Field "key" is deprecated. Use something else.',
        );
      });
    });
  });

  describe('only()', () => {
    beforeEach(() => {
      predicate.only();
    });

    it('errors if default value is not the same type', () => {
      predicate.defaultValue = 123;

      expect(() => {
        predicate.only();
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if value doesnt match the default value', () => {
      expect(() => {
        runChecks(predicate, 'bar');
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if value matches default value', () => {
      expect(() => {
        runChecks(predicate, 'foo');
      }).not.toThrow();
    });
  });

  describe('or()', () => {
    beforeEach(() => {
      predicate.or('bar', 'baz');
    });

    it('errors if no keys are defined', () => {
      expect(() => {
        predicate.or();
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if not 1 option is defined', () => {
      expect(() => {
        runChecks(predicate, 'a', { key: 'foo', struct: {} });
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if at least 1 option is defined', () => {
      expect(() => {
        runChecks(predicate, 'a', { key: 'foo', struct: { foo: 'a' } });
      }).not.toThrow();
    });

    it('doesnt error if at least 1 option is defined that isnt the main field', () => {
      expect(() => {
        runChecks(predicate, 'a', { key: 'foo', struct: { bar: 'b' } });
      }).not.toThrow();
    });
  });

  describe('xor()', () => {
    beforeEach(() => {
      predicate.xor('bar', 'baz');
    });

    it('errors if no keys are defined', () => {
      expect(() => {
        predicate.xor();
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if no options are defined', () => {
      expect(() => {
        runChecks(predicate, 'a', { key: 'foo', struct: {} });
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if more than 1 option is defined', () => {
      expect(() => {
        runChecks(predicate, 'a', {
          key: 'foo',
          struct: {
            foo: 'a',
            bar: 'b',
          },
        });
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if only 1 option is defined', () => {
      expect(() => {
        runChecks(predicate, 'a', {
          key: 'foo',
          struct: {
            foo: 'a',
          },
        });
      }).not.toThrow();
    });
  });

  describe('validate()', () => {
    it('can run the predicate by itself', () => {
      expect(() => {
        func().validate(
          // @ts-ignore Allow invalid type
          123,
        );
      }).toThrowErrorMatchingSnapshot();
    });
  });
});

describe('custom()', () => {
  it('returns a predicate', () => {
    expect(custom(() => {}, '')).toBeInstanceOf(Predicate);
  });

  it('sets type and default value', () => {
    const predicate = custom(() => {}, 123);

    expect(predicate.type).toBe('custom');
    expect(predicate.defaultValue).toBe(123);
  });

  it('returns the type alias', () => {
    expect(custom(() => {}, '').typeAlias()).toBe('custom');
  });
});

describe('func()', () => {
  it('returns a predicate', () => {
    expect(func()).toBeInstanceOf(Predicate);
  });

  it('sets type and default value', () => {
    const noop = () => {};
    const predicate = func(noop);

    expect(predicate.type).toBe('function');
    expect(predicate.defaultValue).toBe(noop);
  });

  it('errors if a non-function value is used', () => {
    expect(() => {
      runChecks(
        func(),
        // @ts-ignore Test invalid type
        123,
      );
    }).toThrowErrorMatchingSnapshot();
  });

  it('returns the type alias', () => {
    expect(func().typeAlias()).toBe('function');
  });
});
