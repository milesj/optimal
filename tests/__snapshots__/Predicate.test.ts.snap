// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Predicate and() errors if no keys are defined 1`] = `"AND requires a list of field names."`;

exports[`Predicate and() errors if not all options are defined 1`] = `"All of these fields must be defined: foo, bar, baz"`;

exports[`Predicate checkType() array errors on non-arrays 1`] = `"Invalid field \\"key\\". Must be an array."`;

exports[`Predicate checkType() boolean errors on non-booleans 1`] = `"Invalid field \\"key\\". Must be a boolean."`;

exports[`Predicate checkType() function errors on non-functions 1`] = `"Invalid field \\"key\\". Must be a function."`;

exports[`Predicate checkType() number errors on non-numbers 1`] = `"Invalid field \\"key\\". Must be a number."`;

exports[`Predicate checkType() object errors on arrays 1`] = `"Invalid field \\"key\\". Must be a plain object."`;

exports[`Predicate checkType() object errors on non-objects 1`] = `"Invalid field \\"key\\". Must be a plain object."`;

exports[`Predicate checkType() object errors on nulls 1`] = `"Invalid field \\"key\\". Null is not allowed."`;

exports[`Predicate checkType() string errors on non-strings 1`] = `"Invalid field \\"key\\". Must be a string."`;

exports[`Predicate custom() is passed entire options object 1`] = `"Invalid field \\"key\\". This will error!"`;

exports[`Predicate custom() triggers callback function 1`] = `"Invalid field \\"key\\". This will error!"`;

exports[`Predicate deprecate() errors for empty value 1`] = `"A non-empty string is required for deprecated messages."`;

exports[`Predicate deprecate() errors for non-string value 1`] = `"A non-empty string is required for deprecated messages."`;

exports[`Predicate errors if default value is undefined 1`] = `"A default value for type \\"string\\" is required."`;

exports[`Predicate invariant() errors if condition is false 1`] = `"Failure"`;

exports[`Predicate invariant() includes a class name 1`] = `"Invalid FooBar field \\"foo.bar\\". Failure"`;

exports[`Predicate invariant() includes a class name when no path 1`] = `"FooBar. Failure"`;

exports[`Predicate invariant() includes a file and class name 1`] = `"Invalid FooBar field \\"foo.bar\\" in package.json. Failure"`;

exports[`Predicate invariant() includes a file name 1`] = `"Invalid field \\"foo.bar\\" in package.json. Failure"`;

exports[`Predicate invariant() includes an option path 1`] = `"Invalid field \\"foo.bar\\". Failure"`;

exports[`Predicate message() errors for empty value 1`] = `"A non-empty string is required for custom messages."`;

exports[`Predicate message() errors for non-string value 1`] = `"A non-empty string is required for custom messages."`;

exports[`Predicate nullable() errors if field is undefined 1`] = `"Invalid field \\"key\\". Null is not allowed."`;

exports[`Predicate only() errors if default value is not the same type 1`] = `"Only requires a default string value."`;

exports[`Predicate only() errors if value doesnt match the default value 1`] = `"Invalid field \\"key\\". Value may only be \\"foo\\"."`;

exports[`Predicate or() errors if no keys are defined 1`] = `"OR requires a list of field names."`;

exports[`Predicate or() errors if not 1 option is defined 1`] = `"At least one of these fields must be defined: foo, bar, baz"`;

exports[`Predicate required() errors if field is undefined 1`] = `"Invalid field \\"key\\". Field is required and must be defined."`;

exports[`Predicate run() errors if value passed is defined and predicate is never 1`] = `"Invalid field \\"key\\". Field should never be used."`;

exports[`Predicate run() errors if value passed is null and predicate is non-nullable 1`] = `"Invalid field \\"key\\". Null is not allowed."`;

exports[`Predicate run() errors if value passed is undefined and predicate is required 1`] = `"Invalid field \\"key\\". Field is required and must be defined."`;

exports[`Predicate run() runs default type of check 1`] = `"Invalid field \\"key\\". Must be a string."`;

exports[`Predicate run() uses custom message 1`] = `"Invalid field \\"key\\". Oops, something is broken."`;

exports[`Predicate validate() can run the predicate by itself 1`] = `"Must be a function."`;

exports[`Predicate xor() errors if more than 1 option is defined 1`] = `"Only one of these fields may be defined: foo, bar, baz"`;

exports[`Predicate xor() errors if no keys are defined 1`] = `"XOR requires a list of field names."`;

exports[`Predicate xor() errors if no options are defined 1`] = `"Only one of these fields may be defined: foo, bar, baz"`;

exports[`func() errors if a non-function value is used 1`] = `"Invalid field \\"key\\". Must be a function."`;
